hvac/api/auth_methods/kerberos.py

```diff
@@ -0,0 +1,485 @@
+        :param keytab_path: (optional) Keytab file path
+        :param krb5conf_path: (optional) explicit krb5.conf path, default paths on your platform.
+        :param use_token: if True, uses the token in the response received from the auth request to set the "token"
+            attribute on the :py:meth:`hvac.adapters.Adapter` instance under the _adapter Client attribute.
+        :type use_token: bool
+        :param mount_point: The "path" the method/backend was mounted on.
+        :type mount_point: str | unicode
+        :return: The response of the login_with_user request.
+        :rtype: requests.Response
+        """
+
+        log = logging.getLogger(self.__class__.__name__)
+        # suppress noisy modules
+        gssapi_log = logging.getLogger("requests_gssapi.gssapi_")
+        if gssapi_log.getEffectiveLevel() == logging.NOTSET:
+            gssapi_log.setLevel(logging.CRITICAL)
+
+        try:
+            vault_host = urlparse(self._adapter.base_uri).netloc.split(':')[0]
+        except Exception:
+            log.error(f'unable to parse Vault url: {self._adapter.base_uri}')
+            vault_host = ''
+
+        # if realm is provided, and username is not a SPN, make it fully qualified
+        if '@' not in username and realm:
+            username = f'{username}@{realm}'
+
+        # handle the likely meanings of the service name that was provided.
+        # we explicitly convert "service" into a gssapi.Name() - this prevents gssapi from attempting to guess
+        # various and illogical combinations of default-realm and service-name, avoiding dubious and confusing errors.
+        # Instead, we try the logical combinations, this makes errors and what was attempted more transparent to
+        # the user.
+        if isinstance(service, str):
+            service_name = service
+            service_names = []
+            # handle logic for the user to specify a service of "name/host"
+            if '/' in service and '@' in service:
+                # you can't do this....
+                suffix = service.rsplit('@')[-1]
+                raise ValueError(f'kerberos service="{service}" is invalid \nA host-based service cannot contain a '
+                                 f'[@] realm qualifier.\nYou should remove the suffix: "@{suffix}"\n... or remove the '
+                                 f'[/] (forward slash) host-based service qualifier')
+            if '/' + HOST_TEMPLATE in service:
+                # this is a host-based service with a _HOST template in it.
+                service = service.replace('/' + HOST_TEMPLATE, '@' + vault_host)
+                log.debug(f'translate "service" from "{service_name}" to "{service}" - host-based-service is assumed')
+                service_names.append(gssapi.Name(service, gssapi.NameType.hostbased_service))
+            elif '/' in service:
+                # if "/" appears by itself convert it to a "@" - which is required by gssapi
```

Comment:
I'm putting this comment in a sort of random place even though it applies elsewhere too, but I wanted to say I _love_ the comments in the code for this type of thing, which I think will go a long way in keeping this code maintainable in the long term.

The only thing I might ask is whenever possible, include a link to where requirements like this are defined or can otherwise be looked up in more detail.